src/modules/manual_control/ManualControl.hpp

```diff
@@ -0,0 +1,180 @@
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_platform_common/defines.h>
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/module_params.h>
+#include <px4_platform_common/posix.h>
+#include <px4_platform_common/px4_work_queue/ScheduledWorkItem.hpp>
+#include <lib/hysteresis/hysteresis.h>
+#include <lib/perf/perf_counter.h>
+#include <uORB/topics/manual_control_input.h>
+#include <uORB/topics/manual_control_switches.h>
+#include <uORB/topics/manual_control_setpoint.h>
+#include <uORB/topics/parameter_update.h>
+#include <uORB/Publication.hpp>
+#include <uORB/Subscription.hpp>
+#include <uORB/SubscriptionCallback.hpp>
+#include "ManualControlSelector.hpp"
+
+using namespace time_literals;
+
+namespace manual_control
+{
+
+class MovingDiff
+{
+public:
+	void update(float value, float dt_s)
+	{
+		if (dt_s < 0.0f) {
+			dt_s = 0.0f;
+
+		} else if (dt_s > _time_period_s) {
+			dt_s = _time_period_s;
+		}
+
+		const float new_diff = value - _last_value;
+		_diff = new_diff * dt_s + _diff * (_time_period_s - dt_s);
+		_last_value = value;
+	}
+
+	float diff() const
+	{
+		return _diff;
+	}
+
+private:
+	float _diff{0.0f};
+	float _last_value{0.0f};
+	const float _time_period_s{1.0f};
+};
+
+
+class ManualControl : public ModuleBase<ManualControl>, public ModuleParams, public px4::ScheduledWorkItem
+{
+public:
+	ManualControl();
+	~ManualControl() override;
+
+	/** @see ModuleBase */
+	static int task_spawn(int argc, char *argv[]);
+
+	/** @see ModuleBase */
+	static int custom_command(int argc, char *argv[]);
+
+	/** @see ModuleBase */
+	static int print_usage(const char *reason = nullptr);
+
+	bool init();
+
+	int print_status() override;
+
+private:
+	static constexpr int MAX_MANUAL_INPUT_COUNT = 3;
+
+	enum class ArmingAction {
+		TOGGLE = -1,
+		DISARM = 0,
+		ARM = 1,
+	};
+
+	enum class ArmingOrigin {
+		GESTURE = 1,
+		SWITCH = 2,
+		BUTTON = 3,
+	};
```

Comment:
I agree but I also hate the fact that we don't have type safe enums in uORB and so this is somewhat nicer/safer for this module alone.

Reply:
I'd agree if it was for this module alone but it's the uORB interface which is not typesafe and is used here. The other side in commander has to reinterpret the values with its own hardcoded definition: https://github.com/PX4/PX4-Autopilot/blob/pr-manual_control_selector/src/modules/commander/Commander.cpp#L854-L856

If there was a definition in the message for the arming reasons you would have it common and could e.g. change numbers without adjusting both sides, full-text search for `ARMING_REASON_RC_GESTURE` and immediately see where it's set and distinguished. Also you could just pass it along to the message generation https://github.com/PX4/PX4-Autopilot/blob/pr-manual_control_selector/src/modules/commander/Commander.cpp#L904-L939 instead of handling each case separately.

Reply:
Alright, done.